README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<!-- badges: start -->

[![R build
status](https://github.com/reconhub/earlyR/workflows/R-CMD-check/badge.svg)](https://github.com/reconhub/earlyR/actions)
[![Codecov test
coverage](https://codecov.io/gh/reconhub/earlyR/branch/master/graph/badge.svg)](https://codecov.io/gh/reconhub/earlyR?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/earlyR)](https://CRAN.R-project.org/package=earlyR)
<!-- badges: end -->

# Welcome to the *earlyR* package\!

This package implements simple estimation of infectiousness, as measured
by the reproduction number (R), in the early stages of an outbreak. This
estimation requires:

  - **prior knowledge**: the **serial interval** distribution, defined
    as the *mean* and *standard deviation* of the (Gamma) distribution.
    In general, these parameters are best taken from the literature.

  - **data**: the daily **incidence** of the disease, including **only
    confirmed and probable** cases.

## Installing the package

To install the current stable, CRAN version of the package, type:

``` r
install.packages("earlyR")
```

To benefit from the latest features and bug fixes, install the
development, *github* version of the package using:

``` r
devtools::install_github("reconhub/earlyR")
```

Note that this requires the package *devtools* installed.

# What does it do?

The main features of the package include:

  - **`get_R`**: a function to estimate *R* as well as the force of
    infection over time, from incidence data; output is an object of
    class `earlyR`

  - **`sample_R`**: a function to obtain a sample of likely *R* values

  - **`plot`**: a function to visualise `earlyR` objects (*R* or force
    of infection).

  - **`points`**: a function using `earlyR` objects to add the force of
    infection to an existing plot.

# Resources

## Worked example

This example is a simplified version of the [*introductory
vignette*](http://www.repidemicsconsortium.org/earlyR/articles/earlyR.html)
(see section below), where `earlyR` is used in conjunction with other
packages to assess infectiousness and growth potential of an early Ebola
Virus Disease (EVD) outbreak. Here, we simply illustrate how `earlyR`
can be used for assessing infectiousness based on a few
confirmed/probable cases.

In this example we assume a small outbreak of Ebola Virus Disease (EVD),
for which the serial interval has been previously characterised. We
study a fake outbreak, for which we will quantify infectiousness (R),
and then project future incidence using the package
[*projections*](https://github.com/reconhub/projections).

The fake data we consider consist of confirmed cases with the following
symptom onset dates:

``` r

onset <- as.Date(c("2017-02-04", "2017-02-12", "2017-02-15",
                   "2017-02-23", "2017-03-01", "2017-03-01",
           "2017-03-02", "2017-03-03", "2017-03-03"))        
```

We assume the current date is 21st March. We compute the daily incidence
using the package [*incidence*](https://github.com/reconhub/incidence):

``` r

library(incidence)
today <- as.Date("2017-03-21")
i <- incidence(onset, last_date = today)
i
#> <incidence object>
#> [9 cases from days 2017-02-04 to 2017-03-21]
#> 
#> $counts: matrix with 46 rows and 1 columns
#> $n: 9 cases in total
#> $dates: 46 dates marking the left-side of bins
#> $interval: 1 day
#> $timespan: 46 days
#> $cumulative: FALSE
plot(i, border = "white")
```

<img src="man/figures/README-incidence-1.png" style="display: block; margin: auto;" />

**Note:** It is **very important to make sure that the last days without
cases are included here**. Omitting this information would lead to an
over-estimation of the reproduction number (*R*).

For estimating *R*, we need estimates of the mean and standard deviation
of the serial interval, i.e. the delay between primary and secondary
symptom onset dates. This has been quantified durin the West African EVD
outbreak (WHO Ebola Response Team (2014) NEJM 371:1481–1495):

``` r

mu <- 15.3 # mean in days days
sigma <- 9.3 # standard deviation in days
```

The function `get_R` is then used to estimate the most likely values of
*R*:

``` r

library(earlyR)
library(ggplot2)

res <- get_R(i, si_mean = mu, si_sd = sigma)
res
#> 
#> /// Early estimate of reproduction number (R) //
#>  // class: earlyR, list
#> 
#>  // Maximum-Likelihood estimate of R ($R_ml):
#> [1] 1.041041
#> 
#> 
#>  // $lambda:
#>   NA 0.01838179 0.0273192 0.03514719 0.0414835 0.04623398...
#> 
#>  // $dates:
#> [1] "2017-02-04" "2017-02-05" "2017-02-06" "2017-02-07" "2017-02-08"
#> [6] "2017-02-09"
#> ...
#> 
#>  // $si (serial interval):
#> A discrete distribution
#>   name: gamma
#>   parameters:
#>     shape: 2.70655567117586
#>     scale: 5.65294117647059
plot(res)
```

<img src="man/figures/README-estimate-1.png" style="display: block; margin: auto;" />

The first figure shows the distribution of likely values of *R*, and the
Maximum-Likelihood (ML) estimation. To derive other statistics for this
distribution, we can use `sample_R` to get a large sample of likely *R*
values, and then compute statistics on this sample:

``` r

R_val <- sample_R(res, 1000)
summary(R_val) # basic stats
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>  0.1902  0.9009  1.1211  1.1774  1.4314  2.7427
quantile(R_val) # quartiles
#>        0%       25%       50%       75%      100% 
#> 0.1901902 0.9009009 1.1211211 1.4314314 2.7427427
quantile(R_val, c(0.025, 0.975)) # 95% credibility interval
#>      2.5%     97.5% 
#> 0.5705706 2.0420420
hist(R_val, border = "grey", col = "navy",
     xlab = "Values of R",
     main = "Sample of likely R values")
```

<img src="man/figures/README-samples-1.png" style="display: block; margin: auto;" />

Finally, we can also represent infectiousness over time using:

``` r
plot(res, "lambdas", scale = length(onset) + 1) +
  geom_vline(xintercept = onset, col = "grey", lwd = 1.5) +
  geom_vline(xintercept = today, col = "blue", lty = 2, lwd = 1.5)
#> Warning: Removed 1 rows containing missing values (position_stack).
```

<img src="man/figures/README-lamdbas-1.png" style="display: block; margin: auto;" />

This figure shows the global force of infection over time, with vertical
grey bars indicating the dates of symptom of onset. The dashed blue line
indicates current day. Note that the vertical scale for the bars is
arbitrary, and only represents the relative force of infection.

## Vignettes

\`Currently available vignettes can be accessed from *R* using:

  - `vignette("earlyR")`: an [introduction to
    `earlyR`](http://www.repidemicsconsortium.org/earlyR/articles/earlyR.html)
    using a simulated Ebola Virus Disease (EVD) outbreak; includes
    projections of future incidence using *projections*.

## Websites

A dedicated website is still in development.

## Getting help online

Bug reports and feature requests should be posted on *github* using the
[*issue*](http://github.com/reconhub/earlyR/issues) system. All other
questions should be posted on the **RECON forum**: <br>
<http://www.repidemicsconsortium.org/forum/>

Contributions are welcome via **pull requests**.

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.


NEWS.md
# earlyR 0.0.5

## Bug fixes

- fix issues in the README and vignetted due to move to ggplot2


# earlyR 0.0.4

## New features

- graphs are now using ggplot2




# earlyR 0.0.3

## Bug fixes

- fix issue #11 where the likelihood surface was incorrectly calculated.
  This unfortunately reverts the issue from #4




# earlyR 0.0.2

## New features

- heat colors in the plot of lambdas have been replaced by a constant color


## Bug fixes

- fixed issue 4 relating to low log-likelihood values which resulted in constant
  zeros when reverted back to the likelihood scale





# earlyR 0.0.1

The main features of the package include:

- **`get_R`**: a function to estimate *R* as well as the force of infection over
    time, from incidence data; output is an object of class `earlyR`

- **`sample_R`**: a function to obtain a sample of likely *R* values

- **`plot`**: a function to visualise `earlyR` objects (*R* or force of infection).

- **`points`**: a function using `earlyR` objects to add the force of infection
    to an existing plot.


tests/testthat.R
library(testthat)
library(earlyR)
library(incidence)

test_check("earlyR")


tests/testthat/helpers.R
library(incidence)
library(distcrete)



tests/testthat/test-print.R
context("Test print function")

test_that("Test against reference results", {
    skip_on_cran()
  
    ## simulate basic epicurve
    dat <- c(0, 2, 2, 3, 3, 5, 5, 5, 6, 6, 6, 6)
    i <- incidence(dat)


    ## example with a function for SI
    si <- distcrete("gamma", interval = 1L,
                    shape = 1.5,
                    scale = 2, w = 0)
    R_1 <- get_R(i, si = si)
    
    expect_equal_to_reference(capture.output(print(R_1)),
                              file = "rds/print_1.rds")
    
})




R/sample_R.R
#' Get a sample of plausible Reproduction Numbers
#'
#' This function derives a sample of plausible R values from an \code{earlyR}
#' object (as returned by \code{\link{get_R}}). The probability of each returned
#' values of R are directly proportional to their likelihood.
#'
#'
#' @author Thibaut Jombart \email{thibautjombart@@gmail.com}
#'
#' @export
#'
#' @examples
#'
#' if (require(incidence)) {
#'  x <- incidence(c(1, 5, 5, 12, 45, 65))
#'  plot(x)
#'  res <- get_R(x, disease = "ebola")
#'  res
#'  plot(res)
#'
#'  sample_R(res, 10)
#'  hist(sample_R(res, 1000), col = "grey", border = "white")
#' }
#'
#' @param x An \code{earlyR} object.
#'
#' @param n The number of R values to sample.
#'

sample_R <- function(x, n = 100) {
  if (!inherits(x, "earlyR")) {
    stop("x is not an earlyR object")
  }

  sample(x$R_grid, prob = x$R_like, n, replace = TRUE)
}


R/internals.R

## ## These function are used to compute infectiousness over time given an
## ## empirical distribution of serial interval.




## ## Get lambda for one individual, one day

## get_lambda_one_day <- function(w, onset, day) {
##   T <- day - onset + 1
##   if (T < 1 || T > length(w)) {
##     return(0)
##   }

##   return(w[T])
## }






## ## Get lambda for one individual, several days.

## get_lambda_days <- function(w, onset, days) {

##   vapply(days,
##          function(day) get_lambda_one_day(w, onset, day),
##          numeric(1))
## }






## ## Get lambda for several individuals, several days.

## get_lambdas <- function(w, onset, days = 1:max(onset)) {
##   if (any(is.na(w))) {
##     stop("NA detected in serial interval")
##   }

##   w <- w/sum(w)
##   all_lambdas <- vapply(seq_along(onset),
##                         function(i)
##                           get_lambda_days(w, onset[i], days),
##                         numeric(length(days)))
##   rowSums(all_lambdas)
## }




## This function converts log-likelihood values back to their original
## scales. It also aims to sanitize very low log-likelihood values which will
## cause numerical approximation problems when converted to the original scale.

loglike_to_density <- function(x) {
  out <- exp(x)
  out <- out / sum(out, na.rm = TRUE)
  out
}


R/plot.earlyR.R
#' Plot earlyR objects
#'
#' These functions are designed for plotting \code{earlyR} objects, output by
#' the function \code{\link{get_R}}. It can plot either the likelihood of R
#' values, or the force of infection over time (see argument \code{type}). For
#' \code{points}, the latter is used.
#'
#' @author Thibaut Jombart \email{thibautjombart@@gmail.com}
#'
#' @export
#'
#' @importFrom ggplot2 .data
#'
#' @rdname plot.earlyR
#'
#' @param x A \code{earlyR} object.
#'
#' @param type The type of graphic to be generated, matching either "R" or
#'   "lamdbas"; "R" will plot the likelihood of R values; "lambdas" will plot
#'   the force of infection over time; see `scale` argument to interprete the
#'   force of infection.
#'
#' @param scale A numeric value indicating the total number of new cases
#'   expected over the time period of the lambdas, or a recognised `character`
#'   string; lambdas will be scaled to correspond to the number of expected
#'   cases every day; defaults to `ml`, which tells function to use the maximum
#'   likelihood estimate of *R* multiplied by the number of infectious cases.
#'
#' @param colour_spectrum Only applied when plotting "lambdas".  If TRUE a
#'   continous spectrum of colours from "red" to "yellow" is used for the bars.
#'
#' @param ... Further arguments to be passed to other methods; for the plot of
#'   *R*, `...` is passed to `ggplot2::geom_line()`; for the plot of *lambdas*,
#'   `...` is passed to `ggplot2::geom_bar()`.
#'
#' @return A `ggplot2` object.
#'
#'
#' if (require(incidence)) {
#'
#' ## example: onsets on days 1, 5, 6 and 12; estimation on day 24
#' onset <- c(1, 5, 6, 12)
#' x <- incidence(onset, last_date = 24)
#' x
#'
#' res <- get_R(x, disease = "ebola")
#' res
#' plot(res)
#' plot(res, "lambdas")
#'
#' }
                                        #
plot.earlyR <- function(x, type = c("R", "lambdas"), scale = "ml",
                        colour_spectrum = TRUE, ...) {

  type <- match.arg(type)

  if (type == "R") {
    df <- data.frame(R = x$R_grid, likelihood = x$R_like)
    df_ml <- data.frame(R_ml = x$R_ml, max_like = max(x$R_like))
    R_ml_label <- paste("R (MLE) =", round(x$R_ml, 2))


    out <- ggplot2::ggplot(df,
                           ggplot2::aes(x = .data$R,
                                        y = .data$likelihood)) +
      ggplot2::geom_line(...) +
      ggplot2::geom_segment(data = df_ml,
                            ggplot2::aes(x = .data$R_ml,
                                         0,
                                         xend = .data$R_ml,
                                         yend = .data$max_like),
                            linetype = 2) +
      ggplot2::geom_label(data = df_ml,
                          ggplot2::aes(x = .data$R_ml,
                                       y = .data$max_like * 1.05,
                                       label = R_ml_label)) +
      ggplot2::labs(x = "reproduction number (R)",
                    title = "Likelihood distribution of R")

  } else {


    ## the scaling value is defined as the sum of all the lambdas weighted by
    ## individual infectiousness; the default value is "ml", in which case the
    ## scaling value is the ML estimate of R multiplied by the number of cases.

    if (tolower(scale) == "ml") {
      scale <- x$R_ml * x$incidence$n
      ylab <- "Force of infection (ML estimate of expected daily cases)"
    } else {
      ylab <- "Force of infection"
    }
    lambdas <- scale * (x$lambdas / sum(x$lambdas, na.rm = TRUE))
    df_plot <- data.frame(date = x$dates, lambda = lambdas)

    if (colour_spectrum) {
      new_col <- grDevices::colorRampPalette(c("red", "yellow"))(length(lambdas))
      out <- ggplot2::ggplot(df_plot,
                             ggplot2::aes(x = .data$date,
                                          y = .data$lambda,
                                          fill = new_col)) +
        ggplot2::scale_fill_manual(values = new_col) +
        ggplot2::guides(fill = FALSE)
    } else {
      out <- ggplot2::ggplot(df_plot,
                             ggplot2::aes(x = .data$date,
                                          y = .data$lambda))
    }
    out <- out +
      ggplot2::theme_bw() +
      ggplot2::geom_bar(stat = "identity", ...) +
      ggplot2::labs(title = "Force of infection over time",
                    x = "date",
                    y = ylab)

  }

  return(out)

}






#' @export
#'
#' @rdname plot.earlyR

points.earlyR <- function(x, scale = 1, ...) {
  plot(x, type = "lambdas", scale = scale, ...)
}


R/print.earlyR.R
#' Print method for earlyR objects
#'
#' This method prints the content of \code{earlyR} objects.
#'
#' @export
#'
#' @author Thibaut Jombart (\email{thibautjombart@@gmail.com})
#'
#' @param x A \code{earlyR} object.
#'
#' @param ... further parameters to be passed to other methods (currently not
#'     used)
#'
print.earlyR <- function(x, ...){
    cat("\n/// Early estimate of reproduction number (R) //")
    cat("\n // class:", paste(class(x), collapse=", "))
    cat("\n")

    cat("\n // Maximum-Likelihood estimate of R ($R_ml):\n")
    print(x$R_ml)
    cat("\n")

    cat("\n // $lambda:\n")
    cat(" ", utils::head(x$lambda))
    if (length(x$lambda)>6L) cat("...")
    cat("\n")

    cat("\n // $dates:\n")
    print(utils::head(x$dates))
    if (length(x$dates)>6L) cat("...")
    cat("\n")

    cat("\n // $si (serial interval):\n")
    print(x$si)

}


R/get_R.R
#' Estimate the Reproduction Number
#'
#' This function estimates the (most of the time, 'basic') reproduction number
#' (R) using i) the known distribution of the Serial Interval (delay between
#' primary to secondary onset) and ii) incidence data.
#'
#' @author Thibaut Jombart \email{thibautjombart@@gmail.com}
#'
#' @export
#'
#' @rdname get_R
#'
#' @param x The daily incidence to be used for inferring the reproduction
#'     number. Input can be an \code{incidence} object, as output by the package
#'     \code{incidence}, or a vector of numbers indicating daily number of
#'     cases. Note that 'zero' incidence should be reported as well (see
#'     details).
#'
#' @param ... Further arguments to be passed to the methods.
#'
#' @details The estimation of R relies on all available incidence data. As such,
#'     all zero incidence after the first case should be included in
#'     \code{x}. When using \code{inidence} from the 'incidence' package, make
#'     sure you use the argument \code{last_date} to indicate where the epicurve
#'     stops, otherwise the curve is stopped after the last case. Use
#'     \code{as.data.frame} to double-check that the epicurve includes the last
#'     'zeros'.
#'
#' @return A list with the \code{earlyR} class, containing the following
#'     components:
#' \itemize{
#' \item \code{$incidence}: the input incidence, in its original format
#'
#' \item \code{$R_grid}: the grid of R values for which the likelihood has been
#' computed.
#'
#' \item \code{$R_like}: the values of likelihood corresponding to the
#' \code{$R_grid}
#'
#' \item \code{$R_ml}: the maximum likelihood estimate of R
#'
#' \item \code{$dates}: the dates for which infectiousness has been computed
#'
#' \item \code{$lambdas}: the corresponding values of force of infection
#'
#' \item \code{$si}: the serial interval, stored as a \code{distcrete} object
#'
#' }
#'
#' @examples
#'
#' if (require(incidence)) {
#'
#' ## example: onsets on days 1, 5, 6 and 12; estimation on day 24
#'  x <- incidence(c(1, 5, 6, 12), last_date = 24)
#'  x
#'  as.data.frame(x)
#'  plot(x)
#'  res <- get_R(x, disease = "ebola")
#'  res
#'  plot(res)
#'
#' }
#'

get_R <- function(x, ...) {
  UseMethod("get_R")
}






#' @export
#' @rdname get_R
#' @aliases get_R.default

get_R.default <- function(x, ...) {
  class <- paste(class(x), collapse = ", ")
  stop("No method for objects of class ", class)
}






#' @export
#' @rdname get_R
#' @aliases get_R.integer

#' @param disease A character string indicating the name of the disease
#'     studied. If provided, then \code{si_mean} and \code{si_sd} will be filled
#'     in automatically using value from the literature. Accepted values are:
#'     "ebola".
#'
#' @param si A \code{distcrete} object (see package \code{distcrete}) containing
#'     the discretized distribution of the serial interval.
#'
#' @param si_mean The mean of the serial interval distribution. Ignored if
#'     \code{si} is provided.
#'
#' @param si_sd The standard deviation of the serial interval
#'     distribution. Ignored if \code{si} is provided.
#'
#' @param max_R The maximum value the reproduction number can take.
#'
#' @param days The number of days after the last incidence date for which the
#'     force of infection should be computed. This does not change the
#'     estimation of the reproduction number, but will affect projections.

get_R.integer <- function(x, disease = NULL, si = NULL,
                          si_mean = NULL, si_sd = NULL,
                          max_R = 10, days = 30, ...) {

  if (sum(x, na.rm = TRUE) == 0) {
    stop("Cannot estimate R with no cases.")
  }
  
  dates <- seq_along(x)
  last_day <- max(dates) + days

  if (is.null(disease)) {
      disease <- "null"
  } else {
      disease <- tolower(disease)
      disease <- match.arg(disease, c("ebola"))
  }


  ## maximum numbers of time steps the distribution of the serial interval is
  ## discretized for

  MAX_T <- 1000


  ## The serial interval is a discretised Gamma distribution. We ensure w(0) = 0
  ## so that a case cannot contribute to its own infectiousness.

  if (is.null(si)) {
      if (is.null(si_mean) && disease == "ebola") {
          si_mean <- 15.3
      }
      if (is.null(si_sd) && disease == "ebola") {
          si_sd <- 9.3
      }
      if (is.null(si_mean)) stop("si_mean is missing")
      if (is.null(si_sd)) stop("si_sd is missing")

      si_param <- epitrix::gamma_mucv2shapescale(si_mean, si_sd / si_mean)

      si_full <- distcrete::distcrete("gamma", shape = si_param$shape,
                                      scale = si_param$scale,
                                      interval = 1L, w = 0L)
  } else {
      if (!inherits(si, "distcrete")) {
          stop("'si' must be a distcrete object")
      }
      if (as.integer(si$interval) != 1L) {
          msg <- sprintf("interval used in si is not 1 day, but %d)",
                         si$interval)
          stop(msg)
      }

      si_full <- si
  }
  si <- si_full$d(seq_len(MAX_T))
  si[1] <- 0
  si <- si / sum(si)

  x_with_tail <- rep(0, last_day)
  x_with_tail[dates] <- x

  
  ## Important note: we cannot compute the likelihood of the first day with
  ## case(s), as the force of infection on that day is by default 0: the
  ## likelihood is in fact conditional on the first day with cases. The
  ## resulting log-likelihood for the first day with cases will be -Inf. We keep
  ## track of this day and ignore this data point when summing over
  ## log-likelihood. In the general case, this will be the first element of x
  ## (x[1]), but we also implement the general case where the incidence curve
  ## could have started before the first cases. Note that the force of infection
  ## for day 1 is also by definition not known (NA), so the first day of data
  ## (x[1]) is always ignored in any case.

  ignored_likelihood <- 1 # fist day always ignored
  if (! sum(x_with_tail) == 0) {
    days_with_cases <- which(x_with_tail > 0)
    first_day_with_cases <- min(days_with_cases, na.rm = TRUE)
    ignored_likelihood <- union(1, first_day_with_cases)
  }
  
  all_lambdas <- EpiEstim::overall_infectivity(x_with_tail, si)
  dates_lambdas <- seq_along(all_lambdas)
  lambdas_for_x <- all_lambdas[dates]

  loglike <- function(R) {
    if (R < 0 ) return(-Inf)
    all_likelihoods <- stats::dpois(x,
                                    lambda = R * lambdas_for_x,
                                    log = TRUE)
    sum(all_likelihoods[-ignored_likelihood])
  }

  GRID_SIZE <- 1000

  R_grid <- seq(0, max_R, length.out = GRID_SIZE)
  R_loglike <- vapply(R_grid, loglike, numeric(1))
  R_like <- loglike_to_density(R_loglike)
  R_ml <- R_grid[which.max(R_like)]

  out <- list(incidence = x,
              R_grid = R_grid,
              R_like = R_like,
              R_ml = R_ml,
              dates = dates_lambdas,
              lambdas = all_lambdas,
              si = si_full)
  class(out) <- c("earlyR", "list")

  return(out)
}






#' @export
#' @rdname get_R
#' @aliases get_R.numeric

get_R.numeric <- function(x, ...) {
    out <- get_R(as.integer(x), ...)
    out$incidence <- x
    return(out)
}






#' @export
#' @rdname get_R
#' @aliases get_R.incidence

get_R.incidence <- function(x, ...) {

  if (as.integer(x$interval) != 1L) {
    msg <- sprintf("daily incidence needed, but interval is %d days",
                   x$interval)
    stop(msg)
  }

  if (ncol(x$counts) > 1L) {
    msg <- "cannot use multiple groups in incidence object"
    stop(msg)
  }

  df <- as.data.frame(x)

  out <- get_R(as.integer(x$counts), ...)

  if (inherits(x$dates, "Date")) {
    out$dates <- min(x$dates) - 1 + out$dates
  }

  out$incidence <- x
  return(out)
}
